core/src/main/scala/wartremover/WartTraverser.scala

```diff
@@ -66,15 +67,35 @@ trait WartTraverser {
     t <:< u.universe.typeOf[Float] ||
     t <:< u.universe.typeOf[Double]
 
-  def hasTypeAscription(u: WartUniverse)(tree: u.universe.ValOrDefDef) : Boolean = {
-    // Correctly handle setters
-    val name = {
-      val n = tree.name.decodedName.toString.trim
-      if(n.endsWith("_=")) s"${n.dropRight(2)}(_=)?"
-      else n
-    }
-    new Regex("""(val|var|def)\s*`?""" + name + """`?(\[.*\])?(\(.*\))*\s*:""")
-        .findFirstIn(tree.pos.lineContent).nonEmpty
+
+  def hasTypeAscription(u: WartUniverse)(tree: u.universe.ValOrDefDef) = {
+    /*
+     * A shameless hack, according to Scala specification Position should be
+     * transparent (isTransparent == true) if it is generated by compiler (either entirely
+     * or partially), but there might be a bug somewhere that tree.tpt.pos.isTransparent
+     * does not work.
+     *
+     * The idea behind hack is simple, when scala compiler generates a tree it simply calls
+     * focus on its owner. focus simply returns the end of the owner's position.
+     */
+    tree.pos.endOrPoint != tree.tpt.pos.startOrPoint
```

Comment:
This is pretty clever (but I wish the macro API didn't make us do stuff like this).